runtime/memory.c

```diff
@@ -729,6 +729,18 @@ CAMLexport caml_stat_string caml_stat_strdup(const char *s)
   return result;
 }
 
+CAMLexport caml_stat_string caml_stat_strndup(const char *s,
+                                              asize_t len, asize_t *out_len)
+{
+  caml_stat_block result = caml_stat_alloc_noexc(len);
```

Comment:
That's not correct. First, you need to clamp at `max(len, strlen(s))`, which in other words is `strnlen(s, len)`. Then you need to add one byte for the `NUL` terminator, and explicitly write at after the `memcpy` call.

Reply:
We can't use strlen - the point here is that the "string" may both contain embedded NUL chars and may even not be NUL-terminated? I believe the code is what is needed (as the dual of the Windows side), but therefore the function name and description may need clarifying/changing (cf. 2nd para of #11449).